lib/spack/spack/test/stage.py

```diff
@@ -728,87 +668,96 @@ def test_first_accessible_path(self, tmpdir):
+        """Test purging of stage directories."""
+        stage_dir = tmpdir.join('stage')
+        stage_path = str(stage_dir)
 
-            # Make sure the cached stage path values are changed appropriately.
-            assert spack.stage._stage_root == path
-        except OSError as e:
-            expected = os.path.join(_non_user_root, getpass.getuser())
-            assert expected in str(e)
+        test_dir = stage_dir.join(path)
+        test_dir.ensure(dir=True)
+        test_path = str(test_dir)
 
-            # Make sure the cached stage path values are unchanged.
-            assert spack.stage._stage_root is None
+        with spack.config.override('config:build_stage', stage_path):
+            stage_root = spack.stage.get_stage_root()
+            assert stage_path == stage_root
 
-    def test_get_stage_root_tmp(self, clear_stage_root, tmp_path_for_stage):
-        """Ensure a temp path stage root is a suitable temp path."""
-        assert spack.stage._stage_root is None
+            spack.stage.purge()
 
-        tmpdir = tmp_path_for_stage
-        path = spack.stage.get_stage_root()
-        assert path == str(tmpdir)
-        assert 'test_get_stage_root_tmp' in path
+            if purged:
+                assert not os.path.exists(test_path)
+            else:
+                assert os.path.exists(test_path)
+                shutil.rmtree(test_path)
 
-        # Make sure the cached stage path values are changed appropriately.
-        assert spack.stage._stage_root == path
+    @pytest.mark.nomockstage
+    def test_get_stage_root_in_spack(self, clear_stage_root):
+        """Ensure an instance path stage root is a suitable path."""
 
-        # Add then purge a few directories
-        dir1 = tmpdir.join('stage-1234567890abcdef1234567890abcdef')
-        dir1.ensure(dir=True)
-        dir2 = tmpdir.join('stage-abcdef12345678900987654321fedcba')
-        dir2.ensure(dir=True)
-        dir3 = tmpdir.join('stage-a1b2c3')
-        dir3.ensure(dir=True)
+        base = canonicalize_path(os.path.join('$spack', '.spack-test-stage'))
+        mkdirp(base)
+        test_path = tempfile.mkdtemp(dir=base)
 
-        spack.stage.purge()
-        assert not os.path.exists(str(dir1))
-        assert not os.path.exists(str(dir2))
-        assert os.path.exists(str(dir3))
+        try:
+            with spack.config.override('config:build_stage',  test_path):
+                path = spack.stage.get_stage_root()
 
-        # Cleanup
-        shutil.rmtree(str(dir3))
+                assert 'spack' in path.split(os.path.sep)
 
-    def test_get_stage_root_in_spack(self, clear_stage_root,
-                                     instance_path_for_stage):
-        """Ensure an instance path stage root is a suitable path."""
-        assert spack.stage._stage_root is None
+                # Make sure cached stage path value was changed appropriately
+                assert spack.stage._stage_root == path
```

Comment:
I think `assert path == test_path` or `assert spack.stage._stage_root == test_path` would be more-straightforward. As-is this is testing that `spack.stage._stage_root == spack.stage.get_stage_root()` (which IMO is less important to test but if you disagree then I'd prefer `assert spack.stage._stage_root == spack.stage.get_stage_root()`).

Reply:
Good point.  It looks like, from the diffs, this was an existing check that got "moved".  I like the second option over the last one since `path` is being used in another assertion at line 750.

Reply:
Changed.